File: python/01-call-endpoint.py
"""
01-call-endpoint.py

Direct x402 call to Hermes Plant's DealAnalyzer (DCF/IRR) endpoint using the
official `x402` Python package. Reads WALLET_PRIVATE_KEY from env.

    WALLET_PRIVATE_KEY=0x... python 01-call-endpoint.py
"""
import json
import os
import sys

from eth_account import Account
from x402.client import X402Client

BASE = os.environ.get("HERMES_BASE_URL", "https://hermesplant.com")
ENDPOINT = os.environ.get("HERMES_ENDPOINT", "/agent-services/dealanalyzer")

pk = os.environ.get("WALLET_PRIVATE_KEY")
if not pk:
    print("Set WALLET_PRIVATE_KEY (0x-prefixed EOA on Base mainnet).", file=sys.stderr)
    sys.exit(2)

account = Account.from_key(pk)
client = X402Client(account=account)

response = client.request(
    "POST",
    f"{BASE}{ENDPOINT}",
    json={
        "cashflows": [-1_000_000, 250_000, 250_000, 300_000, 400_000],
        "discountRate": 0.10,
    },
)

response.raise_for_status()
print("DealAnalyzer result:")
print(json.dumps(response.json(), indent=2))

settlement = response.headers.get("PAYMENT-RESPONSE")
if settlement:
    print("\nSettlement (PAYMENT-RESPONSE header):")
    print(settlement)

File: typescript/src/01-call-endpoint.ts
/**
 * 01-call-endpoint.ts
 *
 * Calls Hermes Plant's DealAnalyzer endpoint (DCF/IRR) using `x402-fetch`.
 * `x402-fetch` is a drop-in replacement for the global `fetch` that handles
 * the 402 → sign → retry handshake automatically.
 *
 *   WALLET_PRIVATE_KEY=0x... npm run example:call
 */
import { createX402Fetch } from "@x402/fetch";
import { privateKeyToAccount } from "viem/accounts";

const BASE = process.env.HERMES_BASE_URL ?? "https://hermesplant.com";
const ENDPOINT = process.env.HERMES_ENDPOINT ?? "/agent-services/dealanalyzer";

const pk = process.env.WALLET_PRIVATE_KEY as `0x${string}` | undefined;
if (!pk) {
  console.error("Set WALLET_PRIVATE_KEY (0x-prefixed EOA on Base mainnet).");
  process.exit(2);
}

const account = privateKeyToAccount(pk);

// x402-fetch wraps the native fetch — when the server returns 402, it
// inspects the PAYMENT-REQUIRED header, signs the appropriate
// PaymentRequirements with `account`, and retries with PAYMENT-SIGNATURE.
const fetch402 = createX402Fetch({ account });

async function main() {
  const response = await fetch402(`${BASE}${ENDPOINT}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      cashflows: [-1_000_000, 250_000, 250_000, 300_000, 400_000],
      discountRate: 0.1,
    }),
  });

  if (!response.ok) {
    console.error(`HTTP ${response.status} ${response.statusText}`);
    console.error(await response.text());
    process.exit(1);
  }

  const result = await response.json();
  console.log("DealAnalyzer result:", JSON.stringify(result, null, 2));

  const settlement = response.headers.get("PAYMENT-RESPONSE");
  if (settlement) {
    const decoded = Buffer.from(settlement, "base64").toString("utf-8");
    console.log("Settlement:", decoded);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
